functions.py

```diff
@@ -0,0 +1,181 @@
+
+import pandas as pd
+import requests
+
+###################
+#### Functions ####
+###################
+
+# function used to read json file
+def read_json(path_file):
+    path = path_file
+    with open(path, "r") as jsonfile:
+        parameters = json.load(jsonfile)
+    return parameters
+
+# function gets encoded utc time and translates it into the format of "%Y-%m-%d, %H:%M:%S"
+def uncode_utc(utc):
+    return time.strftime("%Y-%m-%d, %H:%M:%S", time.localtime(int(utc)))
+
+# semrush updates the reports once a month so the function problematically calculates the date we should have data for
+def mid_month():
+    today = datetime.today().strftime('%Y%m%d')
+
+    today_date = date.today()
+    first = today_date.replace(day=1)
+    lastMonth = first - timedelta(days=1)
+    previous_month = lastMonth.strftime("%Y%m")
+
+    if int(today[
+           6:]) > 20:  # Data is updated two weeks after the end of the month for the previous month, we will have once a month run on the 16th
+        return str(today[:4]) + str(today[4:6]) + str(15)
+    else:
+        return str(previous_month[:4]) + str(previous_month[4:6]) + str(15)
+
+# creates the get request html for the database, separated by the dimension which represents the report type
+def request(dimension, call_type, api_key, features, date_mid_month='null', url='null', domain="null", geo="us",
+            service_url="https://api.semrush.com", phrase="null"):
+    if dimension == 'phrase':
+        params = {
+            'type': call_type,
+            'key': api_key,
+            'phrase': phrase,
+            'database': geo,
+            'export_columns': features,
+            'display_date': date_mid_month,
+        }
+    elif dimension == 'url':
+        params = {
+            'type': call_type,
+            'key': api_key,
+            'target': url,
+            'target_type': 'url',
+            'export_columns': features,
+        }
+    response = requests.get(service_url, params=params)
+    return response
+
+# parse the response from the api, save the results in a list
+def parse_response(call_data):
+    results = []
+    data = call_data.decode('unicode_escape')
+    lines = data.split('\r\n')
+    lines = list(filter(bool, lines))
+    columns = lines[0].split(';')
+    for line in lines[1:]:
+        result = {}
+        for response, datum in enumerate(line.split(';')):
+            result[columns[response]] = datum.strip('"\n\r\t')
+        results.append(result)
+
+    return results
+
+# function makes the actual call to the api, phrase based reports
+# for the phrase based reports the request is done per keyword and per database (each geo has another one)
+# the result of the function is a df
+# as protection the function checks first if we didn't already updated today the tables
+def get_phrase_data(dimension, call_type, features, service_url, api_key, date_mid_month, kw_list, db_list, current_time, last_update_date):

```

Comment:
you don't need to pass a current_time parameter, anywhere you want to use the current time just use datetime.now()